README.md
SPb Python Bot
==============

![spbpython bot status](https://healthchecks.io/badge/67e18dff-5605-4a51-9792-2f8614/REOpd_Jz/spbpython.svg) [![tag](https://img.shields.io/github/tag/nonamenix/spb_python_bot.svg)](./CHANGELOG.md)


The Zen of SPb Python Chat
--------------------------

```python
import __this__
import this
```

[Read Zen of SPb Python Chat](https://github.com/spbpython/orgs-wiki/blob/master/chat/this.md)


Hello world
-----------

```python
import __hello__ 
``` 

Let me google for you
---------------------

### Google

Команда разрешена только для модератора

```
/google query
/g query
```

### Wiki 

Команда разрешена только для модератора

```
/wiki query
/w query
```

PEPs link
---------

Match peps in messages and send links for them.


Environment variables
---------------------

```basg
BOT_TOKEN  # telegram bot token
BOT_LOGGING_LEVEL  # logging level
MODERATORS  # moderators identifiers splited by space
DEBUG  # bot autoreload on file save
```

Run with condo
--------------
[Condo](https://github.com/prepor/condo) — reliable and simple idempotent supervisor for Docker containers.


```edn
{
    :spec {
        :Name "spb_python_bot"
        :Image "nonamenix/spb_python_bot:latest"
        :Env [
            "BOT_TOKEN=339614247:************************************"
            "MODERATORS=132982472 59323058"
            "HEALTHCHECKIO_TOKEN=********-****-****-****-************"
            "BOT_LOGGING_LEVEL=ERROR"
            "MONGO_URL=mongodb://spbpythonbot:Lg2vxjoBT3oRmDd4XnyJ2VD6hJC3jb@ds046677.mlab.com:46677/spb_python_bot"
            "MONGO_HEALTHCHECKIO_TOKEN=fb6c17b8-a64a-41bf-83c7-267ee5124368"
        ]
        :HostConfig {
            :RestartPolicy {:Name "always"}
            :NetworkMode "petprojects_default"
        }
    }

    :health-timeout 45
}
```


CHANGELOG
# Changelog Format:
#   date YYYYMMDD: changes title
#                  changes description
#
# Note: log only changes which go into release
# Note: last records on top
#

Version 0.5 20180121
    * Write all chat updates to mongodb

Version 0.4.2 20180120
    * Add user agent to monitoring

Version 0.4.1 20180120
    * Add monitoring

Version 0.3.7 20170806
    * Fix spaces in Hello world
    * Update python version

Version 0.3.4 20170613
    * Allow import this for all

Version 0.3.3 20170426
    * Fix answer fo non-text message

Version 0.3.2 20170426
    * Match peps with hashtag

Version 0.3.1 20170418
    * Disable web page preview for "let me search for you"

Version 0.3 20170418
    * Moderator commands that allowed only for moderators
    * Let me search for you in google and wiki

Version 0.2: 20170417
    * Zen of Spb Python Chat

Version 0.1: 20170416
    * import this — zen of python
    * import __hello__ - Hello world
    * match peps in text and send link to chat
    * ping-pong
    * Dockerfile


bot/bot.py
from typing import List

from aiotg import Bot as BaseBot, API_TIMEOUT, Chat, asyncio, aiohttp
import motor.motor_asyncio
import re

USER_AGENT = 'SPbPython / 0.4.2'


class Bot(BaseBot):
    healthcheckio_url = 'https://hchk.io/{token}'
    healthcheckio_interval = 150  # seconds
    healthcheckio_token = None

    mongo_healthcheckio_token = None
    mongo_healthcheckio_interval = 1800
    mongo_max_db_size = 512 * 1024 * 1024  # mlab.com free limitation
    mongo_alert_on_fullness = .5
    saved_updates_kinds = ['message', 'edited_message', 'inline_query', 'chosen_inline_result', 'callback_query']

    def __init__(
            self,
            api_token: str, moderators: List[int],
            api_timeout: int = API_TIMEOUT,
            botan_token: str = None,
            healthcheckio_token: str = None,
            mongo_url=None,
            mongo_healthcheckio_token=None,
            name=None):
        super(Bot, self).__init__(
            api_token=api_token,
            api_timeout=api_timeout,
            botan_token=botan_token,
            name=name
        )
        self.healthcheckio_token = healthcheckio_token
        self.moderators = moderators
        self._moderators_commands = []

        if mongo_url is not None:
            mongo_client = motor.motor_asyncio.AsyncIOMotorClient(mongo_url)
            self.mongodb = mongo_client[mongo_url.split('/')[-1]]
            self.mongo_healthcheckio_token = mongo_healthcheckio_token

    async def save_update_to_mongo(self, update):
        collection = list(set(self.saved_updates_kinds).intersection(update.keys()))[0]
        if self.mongodb is not None:
            await self.mongodb[collection].insert_one(update)

    async def still_alive(self):
        async with aiohttp.ClientSession(headers={'User-Agent': USER_AGENT}) as session:
            await session.get(self.healthcheckio_url.format(token=self.healthcheckio_token))

        await asyncio.sleep(self.healthcheckio_interval)
        await self.still_alive()

    async def mongo_still_enough(self):
        stats = await self.mongodb.command('dbstats')
        file_size = stats['fileSize']

        fullness = file_size / self.mongo_max_db_size
        user_agent = 'SPbPython - {}%'.format(fullness * 100)

        if fullness <= self.mongo_alert_on_fullness:
            async with aiohttp.ClientSession(headers={'User-Agent': user_agent}) as session:
                await session.get(self.healthcheckio_url.format(token=self.mongo_healthcheckio_token))

        await asyncio.sleep(self.mongo_healthcheckio_interval)
        await self.mongo_still_enough()

    def run(self, debug=False, reload=None):
        if self.healthcheckio_token or self.mongo_healthcheckio_token:
            loop = asyncio.get_event_loop()

            if self.healthcheckio_token is not None:
                loop.create_task(self.still_alive())

            if self.mongo_healthcheckio_token is not None:
                loop.create_task(self.mongo_still_enough())

        super(Bot, self).run(debug=debug, reload=reload)

    def add_moderators_command(self, regexp, fn):
        """
        Manually register regexp based command
        """
        self._moderators_commands.append((regexp, fn))

    def moderator_command(self, regexp):
        """
        Register a new command that allowed only for moderators

        :param str regexp: Regular expression matching the command to register

        :Example:

        >>> @bot.moderator_command(r"/echo (.+)")
        >>> def echo(chat, match):
        >>>     return chat.reply(match.group(1))
        """

        def decorator(fn):
            self.add_moderators_command(regexp, fn)
            return fn

        return decorator

    def _process_message(self, message):

        chat = Chat.from_message(self, message)

        if message['from']['id'] in self.moderators and 'text' in message:
            for patterns, handler in self._moderators_commands:
                m = re.search(patterns, message["text"], re.I)
                if m:
                    self.track(message, handler.__name__)
                    return handler(chat, m)

        handler = super(Bot, self)._process_message(message)

        return handler

    def _process_update(self, update):
        if self.mongodb is not None:
            loop = asyncio.get_event_loop()
            loop.create_task(self.save_update_to_mongo(update))

        super(Bot, self)._process_update(update)


bot/main.py
import os
from datetime import timedelta, datetime

from urllib.parse import quote

from aiotg import Chat, aiohttp

import logging
from bot import Bot

# Logging
logging.basicConfig(
    level=getattr(logging, os.environ.get('BOT_LOGGING_LEVEL', 'DEBUG')),
    format='%(asctime)s | %(name)s | %(levelname)s - %(message)s'
)
logger = logging.getLogger(__name__)
ch = logging.StreamHandler()
logger.addHandler(ch)

DEBUG = "DEBUG" in os.environ


def get_moderators():
    moderators = [
        132982472,  # nonamenix
        59323058  # lig11
    ]
    try:
        moderators = os.environ['MODERATORS']
    except KeyError:
        pass
    else:
        moderators = [int(m.strip()) for m in moderators.split(' ')]
    logger.info(moderators)
    return moderators


# spb_python_bot
bot = Bot(
    api_token=os.environ["BOT_TOKEN"],
    healthcheckio_token=os.environ["HEALTHCHECKIO_TOKEN"],
    moderators=get_moderators(),
    mongo_url=os.environ.get('MONGO_URL'),
    mongo_healthcheckio_token=os.environ.get('MONGO_HEALTHCHECKIO_TOKEN')
)


def no_more_than_once_every(interval=timedelta(minutes=5), key: str = None):
    restrictions = {}

    def actual_decorator(func):
        def wrapper(*args, **kwargs):
            result = None
            try:
                restriction = restrictions[key]
            except KeyError:
                restrictions[key] = datetime.now() + interval
                result = func(*args, **kwargs)
            else:
                if datetime.now() > restriction:
                    restrictions[key] = datetime.now() + interval
                    result = func(*args, **kwargs)

            return result

        return wrapper

    return actual_decorator


@bot.moderator_command("/?ping")
async def ping(chat, message):
    await chat.reply("pong")


@bot.command("/?import __hello__")
async def hello(chat: Chat, message):
    await chat.reply("Hello world")


@bot.command("/rules")
@bot.command("/?import this")
@bot.command("/?import __this__")
@bot.command("\`import __this__\`")
@no_more_than_once_every(interval=timedelta(minutes=5), key='zen_of_chat')
async def zen(chat: Chat, message):
    # TODO: fetch it from chat_zen_url = "https://raw.githubusercontent.com/spbpython/orgs-wiki/master/chat/this.md"

    await chat.reply("""
*The Zen of SPb Python Chat*
_(Inspired by "The Zen of Python, by Tim Peters")_

- Short introduction of yourself is better than "hello".
- Link to gist is better than source paste.
- One long message is better than many short.
- Editing the message is better than correcting via another one.
- Staying on topic is better than offtopic.
- Good topic is worth discussing though.
- Unless it is started by a link to Habrahabr.
- Politeness counts.
- Bad mood is not a good reason to break the rules.
- Don't ask to ask just ask.
- Text message is better than voice message.
- Unless it is voice conference.
- Git repos are one honking great idea — let's do more of those!
""", parse_mode="Markdown")


def reply_with_let_me_search_for_you(chat: Chat, search_url: str, query: str):
    return chat.send_text(
        search_url.format(query=quote(query)),
        reply_to_message_id=chat.message["message_id"],
        disable_web_page_preview='true',
    )


@bot.moderator_command("/g (?P<query>.+)")
@bot.moderator_command("/google (?P<query>.+)")
async def google(chat: Chat, matched):
    await reply_with_let_me_search_for_you(
        chat,
        search_url="https://www.google.ru/search?q={query}",
        query=matched.group('query')
    )


@bot.moderator_command("/w (?P<query>.+)")
@bot.moderator_command("/wiki (?P<query>.+)")
async def wiki(chat: Chat, matched):
    await reply_with_let_me_search_for_you(
        chat,
        search_url="https://en.wikipedia.org/w/index.php?search={query}",
        query=matched.group('query')
    )


pep_link = "https://www.python.org/dev/peps/pep-{0:04d}/"


async def is_pep_exists(pep):
    async with aiohttp.ClientSession() as session:
        async with session.get(pep_link.format(pep)) as resp:
            return resp.status == 200


@bot.command("\#.*pep-?(?P<pep>\d{1,4})")
async def peps(chat: Chat, matched):
    try:
        pep = int(matched.group('pep'))
    except ValueError:
        pass
    else:
        if await is_pep_exists(pep):
            await chat.send_text(pep_link.format(pep), reply_to_message_id=chat.message["message_id"])


if __name__ == "__main__":
    logger.info("Running...")
    bot.run(debug=DEBUG)
